framework/src/meshgenerators/TransfiniteMeshGenerator.C

```diff
@@ -0,0 +1,542 @@
+  Real rad = computeRadius(P1, P2, P3);
+  Point P0 = computeOrigin(P1, P2, P3);
+
+  // need to shift to center of coordinates to find the corresponding radians
+  Point x0 = (P1 - P0);
+  Point x1 = (P2 - P0);
+
+  // The case when the edge spans quadrants 1 and 4 requires special treament
+  // to periodically switch we compute the angle that needs added to one edge
+  // to identify the entire edge span
+  mooseAssert(x0.norm() > 0.0 && x1.norm() > 0.0,
+              "The point provided cannot generate an arc circle on the edge specified");
+
+  Real arclength = std::acos((x0 * x1) / x0.norm() / x1.norm());
+
+  Real a = std::atan2(x0(1), x0(0));
+  Real b = std::atan2(x1(1), x1(0));
+
+  if (a < 0)
+    a = a + 2 * M_PI;
+  if (std::abs(b - a) > M_PI)
+    b = a + arclength;
+
+  auto it = 0;
+  for (auto rx : param_vec)
+  {
+    Real interval = getMapInterval(rx, 0.0, 1.0, a, b);
+    Real x = P0(0) + rad * std::cos(interval);
+    Real y = P0(1) + rad * std::sin(interval);
+    edge[it] = Point(x, y, 0.0);
+    it++;
+  };
+
+  return edge;
+}
+
+Real
+TransfiniteMeshGenerator::getMapInterval(
+    const Real & xab, const Real & a, const Real & b, const Real & c, const Real & d) const
+// this routine maps a point x\in[a, b] to the corresponding point in the interval [c, d]
+{
+  mooseAssert(std::abs(b - a) > 0.0,
+              "The input interval [a, b] is empty, check that a and b are not identical.");
+  Real xcd = c + (d - c) / (b - a) * (xab - a);
+
+  return xcd;
+}
+
+std::vector<Real>
+TransfiniteMeshGenerator::getPointsDistribution(const Real & edge_length,
+                                                const unsigned int & np,
+                                                const Real & bias) const
+{
+  std::vector<Real> param_vec;
+  Real rx = 0.0;
+
+  if (bias != 1.0)
+  {
+    Real step = 0.0;
+    Real factor = edge_length * (1.0 - std::abs(bias)) / (1.0 - std::pow(std::abs(bias), np - 1));
+    param_vec.push_back(rx);
+    for (unsigned int iter = 1; iter < np; iter++)
+    {
+      step = step + factor * std::pow(bias, Real(iter - 1));
+      rx = getMapInterval(step, 0.0, edge_length, 0.0, 1.0);
+      param_vec.push_back(rx);
+    }
+  }
+  else
+  {
+    const Real interval = edge_length / Real(np - 1);
+    Real rx = 0.0;
+    for (unsigned int iter = 0; iter < np; iter++)
+    {
+      rx = getMapInterval(Real(iter) * interval, 0.0, edge_length, 0.0, 1.0);
+      param_vec.push_back(rx);
+    }
+  }
+  return param_vec;
+}
+
+Real
+TransfiniteMeshGenerator::computeRadius(const Point & P1, const Point & P2, const Point & P3) const
+{
+  Point temp1 = P1 - P2;
+  Real a2 = temp1.norm_sq(); // a is the distance from P1 to P2, but we only it squared
+  Point temp2 = P3 - P1;
+  Real b2 = temp2.norm_sq();
+  Real dr = std::sqrt(b2 - a2 / 4.0);
+  const Real rad = a2 / 8.0 / dr + dr / 2.0;
+  return rad;
+}
+
+Point
+TransfiniteMeshGenerator::computeOrigin(const Point & P1, const Point & P2, const Point & P3) const
```

Comment:
Is there any way to document the algorithm here?  I'm not really following it from the code, and when I plug in some example points (e.g. 3 corners of [-1,1]^2) by hand I hit the "cannot generate an arc circle" case.

Reply:
The algorithm is nothing with an official name, but a mere geometry problem solved via a system of equations, and I added some explanation in the header. I could add a lengthy description of it and the system but felt a tad high-schoolish albeit not very obvious at first.